services/migrations/restore_github.go

```diff
@@ -0,0 +1,915 @@
+func NewGithubExportedDataRestorer(ctx context.Context, githubDataFilePath, owner, repoName string) (*GithubExportedDataRestorer, error) {
+	tmpDir, err := os.MkdirTemp(os.TempDir(), "github_exported_data")
+	if err != nil {
+		return nil, err
+	}
+	// uncompress the file
+	if err := decompressFile(githubDataFilePath, tmpDir); err != nil {
+		return nil, err
+	}
+
+	var restorer = &GithubExportedDataRestorer{
+		ctx:                ctx,
+		githubDataFilePath: githubDataFilePath,
+		tmpDir:             tmpDir,
+		repoOwner:          owner,
+		repoName:           repoName,
+		users:              make(map[string]*githubUser),
+	}
+	if err := restorer.getUsers(); err != nil {
+		return nil, err
+	}
+
+	return restorer, nil
+}
+
+// SetContext set context
+func (r *GithubExportedDataRestorer) SetContext(ctx context.Context) {
+	r.ctx = ctx
+}
+
+// GetRepoInfo returns a repository information
+func (r *GithubExportedDataRestorer) GetRepoInfo() (*base.Repository, error) {
+	type Label struct {
+		URL         string
+		Name        string `json:"name"`
+		Color       string
+		Description string
+		CreatedAt   time.Time
+	}
+	type GithubRepo struct {
+		Name          string `json:"name"`
+		URL           string
+		Owner         string
+		Description   string
+		Private       bool
+		Labels        []Label `json:"labels"`
+		CreatedAt     time.Time
+		DefaultBranch string
+	}
+
+	var githubRepositories []GithubRepo
+	p := filepath.Join(r.tmpDir, "repositories_000001.json")
+	bs, err := ioutil.ReadFile(p)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := json.Unmarshal(bs, &githubRepositories); err != nil {
+		return nil, err
+	}
+	if len(githubRepositories) == 0 {
+		return nil, errors.New("no repository found in the json file: repositories_000001.json")
+	} else if len(githubRepositories) > 1 {
+		return nil, errors.New("only one repository is supported")
+	}
+
+	opts := githubRepositories[0]
+	fields := strings.Split(opts.Owner, "/")
+	owner := fields[len(fields)-1]
+
+	for _, label := range opts.Labels {
+		r.labels = append(r.labels, &base.Label{
+			Name:        label.Name,
+			Color:       label.Color,
+			Description: label.Description,
+		})
+	}
+
+	return &base.Repository{
+		Owner:         r.repoOwner,
+		Name:          r.repoName,
+		IsPrivate:     opts.Private,
+		Description:   opts.Description,
+		OriginalURL:   opts.URL,
+		CloneURL:      filepath.Join(r.tmpDir, "repositories", owner, opts.Name+".git"),
+		DefaultBranch: opts.DefaultBranch,
+	}, nil
+}
+
+// GetTopics return github topics
+func (r *GithubExportedDataRestorer) GetTopics() ([]string, error) {
+	var topics = struct {
+		Topics []string `yaml:"topics"`
+	}{}
+
+	// FIXME: No topic information provided
+
+	return topics.Topics, nil
+}
+
+func (r *GithubExportedDataRestorer) readJSONFiles(filePrefix string, makeF func() interface{}, f func(content interface{}) error) error {
```

Comment:
I don't think `makeF` is a great variable name, what about something like `getJSONType`?

Reply:
I don't think so. This will just create a slice of the object. Not return a type.

Reply:
> I don't think so. This will just create a slice of the object. Not return a type.

Why does it need to be wrapped in a function then?

Reply:
Because I think it's more efficient than use `reflect` package. And do you have any other ideas?

Reply:
> Because I think it's more efficient than use `reflect` package. And do you have any other ideas?

Ah I see, go shenanigans, you cant use it as a direct type. NVM then.